python/ray/data/_internal/datasource/kafka_datasource.py

```diff
@@ -0,0 +1,568 @@
+        """Create read tasks for Kafka partitions.
+
+        Creates one read task per partition.
+        Each task reads from a single partition of a single topic.
+
+        Args:
+            parallelism: This argument is deprecated.
+            per_task_row_limit: Maximum number of rows per read task.
+
+        Returns:
+            List of ReadTask objects, one per partition.
+        """
+
+        # Discover all partitions for all topics
+        # We need to create a consumer on the driver to discover partitions
+        from kafka import KafkaConsumer
+
+        # Build minimal consumer config for partition discovery
+        consumer_config = _build_consumer_config_for_discovery(
+            self.bootstrap_servers, self.authentication
+        )
+
+        # Discover partitions for all topics
+        topic_partitions = []  # List of (topic, partition) tuples
+        discovery_consumer = None
+        try:
+            discovery_consumer = KafkaConsumer(**consumer_config)
+            for topic in self.topics:
+                partitions = discovery_consumer.partitions_for_topic(topic)
+                if not partitions:
+                    raise ValueError(
+                        f"Topic {topic} has no partitions or doesn't exist"
+                    )
+                for partition in partitions:
+                    topic_partitions.append((topic, partition))
+        finally:
+            if discovery_consumer:
+                discovery_consumer.close()
+
+        if not topic_partitions:
+            return []
+
+        # Store config for use in read functions (avoid serialization issues)
+        bootstrap_servers = self.bootstrap_servers
+        start_offset = self.start_offset
+        end_offset = self.end_offset
+        authentication = self.authentication
+        max_records_per_task = self.max_records_per_task
+        timeout_ms = self.timeout_ms
+
+        tasks = []
+        for topic_name, partition_id in topic_partitions:
+
+            def create_kafka_read_fn(
+                topic_name: str = topic_name,
+                partition_id: int = partition_id,
+                bootstrap_servers: List[str] = bootstrap_servers,
+                start_offset: Optional[Union[int, str]] = start_offset,
+                end_offset: Optional[Union[int, str]] = end_offset,
+                authentication: Dict[str, Any] = authentication,
+                timeout_ms: int = timeout_ms,
+            ):
+                """Create a Kafka read function with captured variables.
+
+                This factory function captures configuration variables as default arguments
+                to avoid serialization issues when the read function is executed remotely
+                by Ray. Using default arguments ensures all needed config is available
+                in the remote task without requiring 'self' to be serialized.
+                """
+
+                def kafka_read_fn() -> Iterable[Block]:
```

Comment:
The implementation of this function is pretty long. Wonder if there's a way we can make it shorter (and easier to read) by abstracting some of the logic as functions. 

If you feel like it'd hurt readability though, okay to keep as-is.